internal/controller/resourceset_controller.go

```diff
@@ -296,36 +296,103 @@ func (r *ResourceSetReconciler) checkDependencies(ctx context.Context,
 
 func (r *ResourceSetReconciler) getInputs(ctx context.Context,
 	obj *fluxcdv1.ResourceSet) ([]map[string]any, error) {
-	providers := make([]fluxcdv1.InputProvider, 0)
-	providers = append(providers, obj)
-	for _, inputSource := range obj.Spec.InputsFrom {
-		var provider fluxcdv1.InputProvider
-		key := client.ObjectKey{
-			Namespace: obj.GetNamespace(),
-			Name:      inputSource.Name,
+
+	// inputProviderKey is the key used to identify input providers.
+	type inputProviderKey struct {
+		GVK       schema.GroupVersionKind
+		Name      string
+		Namespace string
+	}
+
+	// getInputProviderKey returns the key for the input provider.
+	getInputProviderKey := func(provider fluxcdv1.InputProvider) inputProviderKey {
+		return inputProviderKey{
+			GVK:       provider.GroupVersionKind(),
+			Name:      provider.GetName(),
+			Namespace: provider.GetNamespace(),
 		}
+	}
+
+	// Initialize the map of providers with the ResourceSet itself.
+	obj.SetGroupVersionKind(fluxcdv1.GroupVersion.WithKind(fluxcdv1.ResourceSetKind))
```

Comment:
Do we really need to mutate the RSET here, it doesn't contain the GVK?

Reply:
Yes, this is always the case when you `.Get()` with a concrete type, the tests fail without this